pkg/didcomm/protocol/introduce/action.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package introduce
+
+import (
+	"sync"
+
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/dispatcher"
+)
+
+// Action errors
+const (
+	ErrChannelRegistered = ActionError("channel is already registered for the action event")
+	ErrNilChannel        = ActionError("cannot pass nil channel")
+	ErrInvalidChannel    = ActionError("invalid channel passed to unregister the action event")
+)
+
+// ActionError defines action error
+type ActionError string
+
+// Error satisfies build-in error interface
+func (e ActionError) Error() string { return string(e) }
+
+// Action definition
+type Action struct {
```

Comment:
@troyronda @soluchok @rolsonquadras 

I think this can be made generic and reused across all state machines. This would remove some of the complexity inside `didexchange.Service`

Reply:
It will be, and also `RegisterMsgEvent ...`

Reply:
Please rebase - PR #444 moved stuff around. `Action` can be moved to the new `didcomm/common/service` package.

As for RegisterMsgEvent, can you also extract those methods into a new type and place it in the new package as well?

Reply:
sure

Reply:
Done!